core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+                The structure of 'value' is:
+                    {
+                        'event_type': str. Type of event.
+                        'session_id': str. ID of the session.
+                    }
+            tuple. For StateHitEventLogEntryModel, a 2-tuple of the form
+                (exp_id, value).
+                The structure of 'value' is:
+                    {
+                        'event_type': str. Type of event.
+                        'version': str. Version of the exploration.
+                        'state_name': str. Name of the state.
+                        'session_id': str. ID of the session.
+                        'created_on': How many milliseconds ago the event was
+                            created.
+                    }
+            tuple. For StateAnswersModel, a 2-tuple of the form
+                (exp_id, value).
+                The structure of 'value' is:
+                    {
+                        'event_type': str. Type of event.
+                        'version': str. Version of the exploration.
+                        'state_name': str. Name of the state.
+                        'submitted_answer_list': List of submitted answers each
+                            of which is stored as a JSON blob.
+                    }
+        """
+        if isinstance(item, stats_models.StartExplorationEventLogEntryModel):
+            yield(
+                item.exploration_id,
+                {'event_type': feconf.EVENT_TYPE_START_EXPLORATION})
+
+        elif isinstance(
+                item, stats_models.CompleteExplorationEventLogEntryModel):
+            yield(
+                item.exploration_id,
+                {
+                    'event_type': feconf.EVENT_TYPE_COMPLETE_EXPLORATION,
+                    'session_id': item.session_id
+                })
+
+        elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+            value = {
+                'event_type': feconf.EVENT_TYPE_STATE_HIT,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'session_id': item.session_id,
+                'created_on': utils.get_time_in_millisecs(item.created_on)
+            }
+            yield(item.exploration_id, value)
+
+        else:
+            value = {
+                'event_type': MigrateStatistics.EVENT_TYPE_STATE_ANSWERS,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'submitted_answer_list': item.submitted_answer_list
+            }
+            yield(item.exploration_id, value)
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        num_starts = 0
+        num_completions = 0
+
+        exp_id = key
+        exploration = exp_services.get_exploration_by_id(exp_id)
+        latest_exp_version = exploration.version
+
+        # Check if model already exists. If it does, delete it.
```

Comment:
Actually, can't you just modify the existing model at the end instead? Two reasons:

(a) If something goes wrong in the code that follows, there'll still be at least some data shown in the UI. The model also only gets updated when the computation below is complete; there isn't this interim phase of "no model available".
(b) A single put() is faster than a delete() + create().